Add OIDC method to Kafka authentication

richscott opened the pull request:
This change adds support for authentication via OIDC to the Kafka client. It provides an  implementation of the sarama.AccessTokenProvider interface, supporting the `client_credentials` Grant Type, and a background token refresh.

richscott committed:
WIP - add OIDC support to Kafka auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
WIP - add OIDC support to Kafka auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Move OIDC client and test into sub-package.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'feature/kafa-openid-auth' of github.com:richscott/opentelemetry-collector-contrib into feature/kafa-openid-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Start new implementation of OIDC server.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
golangci-lint fixes; fixes to use new OIDC server

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Get OIDC mock server goroutine working.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Reduce wait for mock server startup to 100ms.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Set our expiresIn from either token's ExpiresIn or Expiry

Add more assertions on the inner token string in the resultant
sarama access token.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafa-openid-auth

richscott committed:
Simplify mock OIDC server invocation args; remove dead code

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Use dynamic port allocation for mock OIDC server

This is so we can quickly stop/start a new server in each
test func.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Add test to verify token-caching.

Also give some vars better names; disable some logging
messages.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Add test for expired token; lint fixes

Remove debugging messages.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafa-openid-auth

richscott committed:
Tidy up go.mod and go.sum

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove global expireSecs, pass down into handler and generator

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Start background refresher when OIDC provider is created

Also, manually build a (small subset of a) K8S token, for unit-testing.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafa-openid-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafa-openid-auth

richscott committed:
Use a var for token expire; add refreshAhead test

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Add more assertions for refreshAhead test.

Also accept and propagate EndpointParams and AuthStyle
fields for oauth2 clientcredentials.Config.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafa-openid-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Add wiring struct and logic for configureSASL()

Also add authentication configuration setup test.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Remove old unused MockOauthProvider type and member func

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'feature/kafka-oidc-auth' of github.com:richscott/opentelemetry-collector-contrib into feature/kafka-oidc-auth

richscott committed:
Fix missing oauth2 module in kafkametricsreceiver package

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Un-camelcase variable name, per golangci-lint.

Also add outh2 mention in go.mod/go.sum for kafkareceiver.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Add oauth2 module to Kafka exporter and observer

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Use separate err var for server shutdown.

We cannot use the same, earlier `err` variable, as it
causes a sporadic race condition.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Add cancellation of background token refresher goroutine

This avoids leaking the refresher goroutine, notably when running
the unit tests.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Add initial draft of change log notes.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Finish change log entries.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove debugging messages.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

MovieStoreGuy commented:
Hey @richscott ,

Please make sure you sign the CLA so we can include the changes into the project.

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Fix formatting warning from golangci-lint.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Set and use refreshCoolDown on OIDC token provider.

Also check if our context is done before calling updateToken().

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Fix go.mod for kafkareceiver

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Rewrite and simplify OIDC token background refresher logic.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Use t.Context() in kafka OIDC client tests.

Signed-off-by: Rich Scott <richscott@sent.com>

atoulme commented:
CLA is signed, getting reviews in.

axw reviewed (commented):
@richscott thanks for the PR. I've skimmed the code, and the code looks good.

My main question at the moment is: does all of this OAuth2 logic have live in the Kafka components? There appears to be a significant overlap with what already exists in [oauth2clientauthextension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/oauth2clientauthextension).

I think what would be ideal is if the Kafka components could be configured with the SASL/OAUTHBEARER mechanism, and the configuration is just a reference to an instance of oauth2clientauthextension. oauth2clientauthextension would need to be modified to expose an API for the Kafka client code to invoke to obtain a token.

Aside from that, franz-go is about to become the default and in future only client implementation. I would recommend that we start with franz-go, and perhaps not even bother with Sarama.

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on .chloggen/feature_kafka-oidc-auth.yaml:
```diff
@@ -0,0 +1,29 @@
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: |
+  It provides an implementation of the sarama.AccessTokenProvider interface, supporting the
+  client_credentials Grant Type, and a background token refresh.
```
@axw Thanks, and I agree completely with all your points/questions. This whole effort was animated by a request from another team of my employer, and they suggested this implementation to extend the Sarama client mechanism to do OIDC, and notably, to get the client_secret from a file (in usage scenario of using a Kubernetes pod secret file, which is dynamic and rotated frequently). I see now that the `oauth2clientextension` already (very closely) has this support, including `client_id_file` and `client_secret_file`. (Full disclosure: I'm new to this project/repo, and working with OIDC as well.) Also, as the franz-go client migration will occur soon(ish), I'll go ahead and change this code to use that in lieu of Sarama. I think I can excise a lot of this, work through the call-chain from the franz_client.go constructors, and then modify `configkafka.SASLConfig` and its users to use the oauth2clientextension.

axw replied:
Thanks @richscott! That sounds good. I'll move this PR back to draft in the mean time.

axw replied:
@richscott thanks for the changes. I'm glad to see most of the logic now lives in oauth2clientauthextension. I'd like to see if we can go even further.

What I have in mind is that users will configure `oauth2clientauthextension` independently of the Kafka components, and then reference it via a new config attribute in the Kafka components. Something like this:

```yaml
extensions:
  oauth2client:
    client_id_file: /path/to/client_id_file
    client_secret: /path/to/client_secret_file

exporters:
  kafka:
    auth:
      sasl:
        mechanism: OAUTHBEARER
        oauthbearer_token_source: oauth2client
```

The "oauthbearer_token_source" config setting would be the ID of an extension that implements a token source interface along these lines:

```go
import "golang.org/x/oauth2"

type TokenSource interface {
    Token(context.Context) (*oauth2.Token, error)
}
```

This way we're not duplicating any oauth2 config, and would be closer to how the oauth2client extension works with HTTP/gRPC-based exporters -- see https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configauth/README.md

axw replied:
If it helps, we can just skip the Sarama implementations and say that this functionality is only supported when using franz-go. The Sarama code has been removed from the exporter already, and should eventually be removed from kafkametrics receiver too.

axw replied:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/45336 injects the host for franz-go

richscott commented on internal/kafka/franz_client.go:
```diff
@@ -32,6 +32,7 @@ const (
 	SCRAMSHA256          = "SCRAM-SHA-256"
 	PLAIN                = "PLAIN"
 	AWSMSKIAMOAUTHBEARER = "AWS_MSK_IAM_OAUTHBEARER" //nolint:gosec // These aren't credentials.
+	OIDCFILE             = "OIDCFILE"
```
This was a early-development mistake - I'll be definitely pulling this out.

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Migrate OIDC file-based token auth to use franz client

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Rework if/else to switch to fix lint error.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Replace logic in internal/kafka/oidc/* with existing oauth2 extension

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Update description of NewOIDCTokenProvider func.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Move internal/kafka/oidc/* to internal/kafka

There's only a single program file and its test file in the former
sub-dir, and adding a go.mod to the dir just to import the
oauth2authclientextension package seemed more complicated than just
having both oidc_client*.go reside in internal/kafka directly.

Also added a 'replace' directive in the kafkareceiver/go.mod, so
oidc_client can get the latest Config change in
oauth2authclientextension before all this is published. Unit tests
in receiver/kafkareceiver and internal/kafka are all successful (as
well as the rest of the repo).

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
go.mod fixes for each kafka component

Add a 'replace' rule so that each kafka component's go.mod uses the
direct source in extension/oauth2clientauthextension, so it has the
latest (e.g. this PR branch) changes available, so the OIDC client logic
available to each kafka component can find the Config definition in
oauth2clientauthextension.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Correct changelog component entry.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

MovieStoreGuy commented on internal/kafka/franz_client.go:
```diff
@@ -318,6 +319,19 @@ func configureKgoSASL(cfg *configkafka.SASLConfig) (kgo.Opt, error) {
+				cfg.OIDCFILE.ClientSecretFilePath, cfg.OIDCFILE.TokenURL,
+				cfg.OIDCFILE.Scopes, cfg.OIDCFILE.EndPointParams,
+				cfg.OIDCFILE.AuthStyle, cfg.OIDCFILE.ExpiryBuffer)
+			_ = cancel // Store cancel function for cleanup if needed
```
This is just assigning the cancel function to the blank operator, did you mean to do something more with it?

richscott replied:
Good catch - I originally thought perhaps it might be needed/useful for a context cancellation invocation, but that's not done anywhere in this call-chain on upward - removed in commit f1d88f04ae .

MovieStoreGuy commented on pkg/kafka/configkafka/config.go:
```diff
@@ -363,18 +365,23 @@ type SASLConfig struct {
 	case "AWS_MSK_IAM_OAUTHBEARER":
 		// TODO validate c.AWSMSK
+	case "OIDCFILE":
+		// TODO valid c.OIDCFile
```
This should be done as part of the change, not after since it can be partially released to users.

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Rename OIDCFILE to OAUTHBEARER; implement validation logic.

Rename "OIDCFILE" mechanism to "OAUTHBEARER", per review suggestion.
Implement basic validation for OAUTHBEARER. Add guard against unkeyed
struct initialization.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove unused context cancel var.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Fix format specifier on error message.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

atoulme commented:
Please resolve conflicts and mark ready to review again.

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Remove OAUTHBEARER config structs and logic in lieu of using extension

Expect the oauth2clientauthextension to be co-resident and use that
for getting a TokenProvider for kafka authentication. Remove all local
OAUTHBEARER configuration in kafka, and get the configured/running
TokenProvider from the injected component.Host map.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

axw commented on internal/kafka/oidc_client.go:
I think this file can be removed now, right?

axw reviewed (commented):
Thanks @richscott, I think this is nearly there.

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Remove now-unnecessary oidc_client.go and use of NewOIDCTokenProvider().

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Replace hard-wired extension name in configuration check.

In configureKgoSASL(), replace hard-wired "oauth2client" component ID
name with new configkafka.SASLConfig member OauthBearerTokenSource;
validate type of component and check for name existence.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

axw commented:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/45917

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Remove now-redundant ClientCredentialsConfig struct

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Start using new oauth2clientauthextension.ContextTokenSource.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Add details in change entry on oauth2 configuration

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Update tidylist.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Update config schema for kafka config.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott commented:
Hi, @MovieStoreGuy or @axw - I fixed the tidylist and config schema checks and refreshed from `main` so this branch should be current - can the CI workflows be run again so I can see if there are any remaining issues?
Thanks.

axw commented on pkg/kafka/configkafka/config.go:
```diff
@@ -424,6 +429,11 @@ func (c SASLConfig) Validate() error {
 		}
+	case "OAUTHBEARER":
+		if c.OAuthBearerTokenSource.Type().String() != component.KindExtension.String() ||
+			c.OAuthBearerTokenSource.Name() == "" {
```
```suggestion
		if c.OAuthBearerTokenSource == (component.ID{}) {
```

`Type` will be something like "oauth2client", see https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/72970c03ea7ef45d7b40e26a9981fbfb9b9f26d0/extension/oauth2clientauthextension/internal/metadata/generated_status.go#L10

`Name` is optional. It'll be empty if you configure an extension just as "oauth2client", but you can also create named extensions like "oauth2client/google" or "oauth2client/github" for example.

axw commented on internal/kafka/franz_client.go:
```diff
@@ -28,6 +28,7 @@ import (
 	"go.opentelemetry.io/collector/config/configcompression"
 	"go.uber.org/zap"
 
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/oauth2clientauthextension"
```
Let's remove the direct dependency on that extension, and create a local copy of the interface:

```go
type contextTokenSource interface {
        Token(context.Context) (*oauth2.Token, error)
}
```

Later on we might want to add an integration test, but we can worry about that later.

axw reviewed (commented):
Thanks @richscott, just a few more simplifications.

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Incorporate suggestions from PR review.

Signed-off-by: Rich Scott <richscott@sent.com>

richscott commented on internal/kafka/franz_client.go:
```diff
@@ -300,6 +302,22 @@ func configureKgoSASL(cfg *configkafka.SASLConfig) (kgo.Opt, error) {
+		extMap := host.GetExtensions()
+		oauthExt, exists := extMap[cfg.OAuthBearerTokenSource]
+		if !exists {
+			return nil, fmt.Errorf("extension %s is not configured", cfg.OAuthBearerTokenSource.Name())
```
Fixed in commit a5d98cbf5d.

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

axw reviewed (commented):
Thanks @richscott, looks like the go.mod files need an update but otherwise LGTM.

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Remove extraneous import in go.mod; fix ordering in franz_client.go

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Remove extranous oauth2clientauthextension go.mod mentions

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Signed-off-by: Rich Scott <richscott@sent.com>

richscott committed:
Merge branch 'main' into feature/kafka-oidc-auth

Commit:
Merge branch 'main' into feature/kafka-oidc-auth

Commit:
Fix tidylist

axw reviewed (approved):
Thanks for your patience and persistence, @richscott. LGTM.

paulojmdias commented:
/workflow-approve
